File: analysis/rpc/bugs.go
// Copyright 2022 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package rpc

import (
	"fmt"

	"go.chromium.org/luci/analysis/internal/bugs"
	configpb "go.chromium.org/luci/analysis/proto/config"
	pb "go.chromium.org/luci/analysis/proto/v1"
)

func createAssociatedBugPB(b bugs.BugID, cfg *configpb.ProjectConfig) *pb.AssociatedBug {
	// Fallback bug name and URL.
	linkText := fmt.Sprintf("%s/%s", b.System, b.ID)
	url := ""

	switch b.System {
	case bugs.MonorailSystem:
		monorailCfg := cfg.BugManagement.GetMonorail()
		if monorailCfg == nil {
			// Try fetching from legacy config location.
			monorailCfg = cfg.Monorail
		}
		if monorailCfg == nil {
			break
		}
		project, id, err := b.MonorailProjectAndID()
		if err != nil {
			// Fallback to basic name and blank URL.
			break
		}
		if project == monorailCfg.Project {
			if monorailCfg.DisplayPrefix != "" {
				linkText = fmt.Sprintf("%s/%s", monorailCfg.DisplayPrefix, id)
			} else {
				linkText = id
			}
		}
		if monorailCfg.MonorailHostname != "" {
			url = fmt.Sprintf("https://%s/p/%s/issues/detail?id=%s", monorailCfg.MonorailHostname, project, id)
		}
	case bugs.BuganizerSystem:
		linkText = fmt.Sprintf("b/%s", b.ID)
		url = fmt.Sprintf("https://issuetracker.google.com/issues/%s", b.ID)
	default:
		// Fallback.
	}
	return &pb.AssociatedBug{
		System:   b.System,
		Id:       b.ID,
		LinkText: linkText,
		Url:      url,
	}
}

File: milo/ui/src/monitoringv2/hooks/bugs.ts
// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { uniq } from 'lodash-es';
import { useEffect, useMemo } from 'react';

import { useInfiniteIssueListQuery } from '@/common/hooks/gapi_query/corp_issuetracker';
import { Bug, bugFromJson } from '@/monitoringv2/util/server_json';

import { useTree } from '../pages/monitoring_page/context';

import { useAlertGroups } from './alert_groups';

export interface BugsData {
  otherHotlistBugs: Bug[];
  groupBugs: { [groupName: string]: Bug[] }; // Bugs associated with alert groups.
  bugQueryError: unknown;
  isBugQueryError: boolean;
  bugIsRefetchError: boolean;
  bugIsLoading: boolean;
}

export function useBugs(): BugsData {
  const tree = useTree();
  const { alertGroups } = useAlertGroups();

  const linkedBugs = uniq(alertGroups.flatMap((g) => g.bugs));

  const {
    hasNextPage: bugHasNextPage,
    fetchNextPage: bugFetchNextPage,
    data: bugData,
    error: bugQueryError,
    isError: isBugQueryError,
    isRefetchError: bugIsRefetchError,
    isLoading: bugIsLoading,
  } = useInfiniteIssueListQuery(
    {
      query: `(status:open AND hotlistid:${tree?.hotlistId})${
        linkedBugs.length > 0 ? ' OR ' : ''
      }${linkedBugs.map((b) => 'id:' + b).join(' OR ')}`,
      orderBy: 'priority',
      pageSize: 50,
    },
    {
      refetchInterval: 60000,
      enabled: !!tree?.hotlistId,
    },
  );

  useEffect(() => {
    if (bugHasNextPage) {
      bugFetchNextPage();
    }
  }, [bugFetchNextPage, bugHasNextPage, bugData?.pages.length]);

  const bugs = useMemo(
    () =>
      bugData?.pages.flatMap((p) => p.issues || []).map((i) => bugFromJson(i)),
    [bugData],
  );

  const hotlistBugs = useMemo(
    () =>
      bugs?.filter((b) =>
        alertGroups.every((g) => !g.bugs.includes(b.number)),
      ) || [],
    [bugs, alertGroups],
  );

  const groupBugs = useMemo(() => {
    const result: { [groupName: string]: Bug[] } = {};
    for (const group of alertGroups) {
      result[group.name] =
        bugs?.filter((b) => group.bugs.includes(b.number)) || [];
    }
    return result;
  }, [alertGroups, bugs]);

  return {
    otherHotlistBugs: hotlistBugs,
    groupBugs,
    bugQueryError,
    isBugQueryError,
    bugIsRefetchError,
    bugIsLoading,
  };
}
